#DP-83 feat: Profile Settings

samyshyaka opened the pull request:
#### What does this PR do?
Enables a user to update their profile settings

#### Description of the tasks to be completed
- [x] Users should be able to update their personal information
- [x] Users should be able to update their avatar
 
#### How should this be manually tested?
- Run the commands below in your terminal
```js
git checkout ft-profile-settings-dp-83
git fetch
npm install
npm run dev
open http://localhost:4000/
```
- Head over to your Appolo Studio
- Send a request to create a user; when you create the user, their profile is created as well. From the response you can retrieve their profile id (Refer to screenshot 1)
- Use the retrieved profile id to target the specific profile and update it ((Refer to screenshot 2) 
NB: Remember to set the authorization variable with the provided token, to be able to update the user's profile 

#### Pre-requisites:

- Clone the project

#### Relevant trello story:
[#DP-83](https://trello.com/c/X247Bww7/83-profile-settings)

#### Screenshots (if appropriate)

1. Create user

![create_user](https://user-images.githubusercontent.com/84643891/185812816-00604422-0af9-4b90-b438-00bd30ddfbb1.png)

2. Update user's profile

![update_profile](https://user-images.githubusercontent.com/84643891/185812845-18e128e5-e348-435d-ac9a-08d6a056af19.png)

jeandamascyifuzo reviewed (approved):
well done @samyshyaka

feyton commented on src/models/user.ts:
```diff
@@ -15,7 +15,7 @@ const userSchema = new Schema({
         default: 'user',
     },
-    profile: {
+    profile_id: {
```
This field should remain as `profile` because when you call the user with the `populate` option, it is more meaningful by saying 
```
// get user with profile would be easier
const user = await User.findOne({where: {id:id}}).populate("profile", ["first_name",...])

```
Since we are not just storing the id despite the field type, we are referencing another object which is `profile` not `profile_id`

feyton commented on src/models/user.ts:
```diff
@@ -32,23 +32,39 @@ userSchema.methods.checkPass = async function (password:string) {
     'Profile',
     new Schema({
-        firstName: {
+        email: {
```
Since each user will have their emails which is not editable,

You can safely remove this field since it would mean requiring keeping in sync two fields on two different models that are already linked.

feyton commented on src/models/user.ts:
```diff
@@ -32,23 +32,39 @@ userSchema.methods.checkPass = async function (password:string) {
             type: mongoose.Types.ObjectId,
             ref: 'User',
-            required: true,
             unique: true,
```
Make this field `required` to avoid creating `profiles` that are not linked to any existing user. And we can also take advantage of the `mongoose` events triggers to do database cleanups. This will requires a change in models definition where we separate the models from schemas (At least to my knowledge)

Example

```
// Delete related profile when a user is deleted to keep a clean relationship
const userSchema = new Schema({

// add all existing fields definitions
})
userSchema.pre("remove", async function (){
await Profile.findAndRemove({where: {_id: this.profile})
})
// this solution can be customized, even be used to automatically create a profile on user creation
// you can use userSchem.post("save", callback)

```

feyton commented on src/resolvers/profileResolver.ts:
```diff
@@ -0,0 +1,53 @@
+import cloudinary from 'cloudinary';
+import path from 'path';
+
+cloudinary.config({
```
Remember to add these values in the `.env.example`, I mean the names, not actually credentials

feyton commented on src/resolvers/profileResolver.ts:
```diff
@@ -0,0 +1,53 @@
+    Mutation: {
+        updateProfile: async (parent:any, args:any, context: any) => {
+            try {
+              let { _id, firstName, lastName, address, city, country, phoneNumber, biography, fileName } = args
```
reduce the use of `let` to favor `const` as the ES6 standard to avoid giving large scope to the variables that will not be `mutated` afterwards

feyton commented on src/resolvers/profileResolver.ts:
```diff
@@ -0,0 +1,53 @@
+                country: country,
+                phoneNumber: phoneNumber,
+                biography: biography,
+                avatar: `${photo?.public_id}.${photo?.format}`
```
I believe cloudinary returns a `url` automatically from the response. To ease the integration, save the `url` to the uploaded picture. 
That way, on frontend we wont have to deal with `URL building` which is unnecessary.

```
avatar: photo.url
// the return is already a string and template literals are not needed
```

feyton commented on src/schema/index.ts:
```diff
@@ -1,12 +1,12 @@
     email: String!
     password: String!
-    profile: Profile!
+    profile_id: String!
```
This is not just an id, but a children of the user schema that is resolved below.
You can keep the name `profile` to avoid confusion

feyton commented on src/schema/index.ts:
```diff
@@ -18,17 +18,21 @@ const Schema = gql`
   }
   type Profile {
-    id: ID!
+    _id: ID!
```
Also, here  I would argue that the `schema`should retain the `id` field not the `_id` which is in reference to mongoose style of having the prefix on the id.

Though this would be a matter of preference, so feel free to resolve it.

feyton reviewed (changes requested):
### Requesting changes
@samyshyaka , 
I have left some comments for the changes that you can make for the pull request. Please refer to these changes mentioned below too

- Why do we have uploaded images even though we are using cloudinary as our media server which require not file on our server. You can delete them after they have been uploaded, or we can use `buffer` or `BASE64` to upload which would require no cleanup afterwards
- Also remember to share new environment variable to ease testing

feyton commented on src/resolvers/profileResolver.ts:
```diff
@@ -0,0 +1,54 @@
+                    throw new Error("You need to login first")
+                }
+
+                if (fileName)  {
```
The variable that is defined here will be available for the following block

Referring to the `photo` which you have also referenced below, which will raise an error of `undefined` due to scoped nature of constant

feyton commented on src/resolvers/profileResolver.ts:
```diff
@@ -0,0 +1,54 @@
+                country: country,
+                phoneNumber: phoneNumber,
+                biography: biography,
+                avatar: photo?.secure_url
```
you will not have access to the `photo` variable since it was defined in the `if` block which tie the `photo` to that block only.

Please adjust

feyton reviewed (changes requested):
@samyshyaka 
This is really great work. Please address one issue that I mentioned to cater for the block error. Others things are well implemented.

I will take a look into how we can improve file upload

feyton commented on src/schema/index.ts:
```diff
@@ -22,13 +22,16 @@ const Schema = gql`
-  }
-  type Query {
     getAllProfiles: [Profile]
     getProfile(id: ID!): Profile
```
@samyshyaka 
There is no need for the ID since we can get the user from the token and use to get the profile

feyton reviewed (changes requested):
@samyshyaka 
Look into the comment I just left

chanceDira reviewed (approved):
Good job @samyshyaka

jeandamascyifuzo reviewed (approved):
well done @samyshyaka

ceelogre reviewed (commented):
Let's first have the integration to test this.

ceelogre reviewed (dismissed):
There's virtually no use of typescript language in your feature. It's just js saved in a ts file. Find a way to use ts language features.

rukundo-kevin commented on src/schema/index.ts:
```diff
@@ -22,15 +22,17 @@ const Schema = gql`
+    country: String    
+    phoneNumber: String
+    biography: String
+    avatar: String
```
Can you add a coverImage as it is required in the frontend.

ceelogre reviewed (commented):
Resolve all comments raised by reviewers, add instructions on how to test using the frontend app and all fix status checks.

Procech20 committed:
feat(profile): implement profile settings

- User is able to update their profile info
- The user is able to update their avatar

feyton commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Profile Settings Integration](https://trello.com/c/X4G3QkHh/103-profile-settings-integration)

Dev-nkundineza reviewed (changes requested):
Good work
